File: managed/src/SwiftlyS2.Generated/Natives/Hooks.cs
#pragma warning disable CS0649
#pragma warning disable CS0169

using System.Text;
using System.Threading;
using SwiftlyS2.Shared.Natives;

namespace SwiftlyS2.Core.Natives;

internal static class NativeHooks
{
    private static int _MainThreadID;

    private unsafe static delegate* unmanaged<nint> _AllocateHook;

    public unsafe static nint AllocateHook()
    {
        var ret = _AllocateHook();
        return ret;
    }

    private unsafe static delegate* unmanaged<nint> _AllocateVHook;

    public unsafe static nint AllocateVHook()
    {
        var ret = _AllocateVHook();
        return ret;
    }

    private unsafe static delegate* unmanaged<nint> _AllocateMHook;

    public unsafe static nint AllocateMHook()
    {
        var ret = _AllocateMHook();
        return ret;
    }

    private unsafe static delegate* unmanaged<nint, void> _DeallocateHook;

    public unsafe static void DeallocateHook(nint hook)
    {
        _DeallocateHook(hook);
    }

    private unsafe static delegate* unmanaged<nint, void> _DeallocateVHook;

    public unsafe static void DeallocateVHook(nint hook)
    {
        _DeallocateVHook(hook);
    }

    private unsafe static delegate* unmanaged<nint, void> _DeallocateMHook;

    public unsafe static void DeallocateMHook(nint hook)
    {
        _DeallocateMHook(hook);
    }

    private unsafe static delegate* unmanaged<nint, nint, nint, void> _SetHook;

    /// <summary>
    /// the callback should receive the exact arguments as the function has, and to return the same amount of arguments
    /// </summary>
    public unsafe static void SetHook(nint hook, nint func, nint callback)
    {
        _SetHook(hook, func, callback);
    }

    private unsafe static delegate* unmanaged<nint, nint, int, nint, byte, void> _SetVHook;

    /// <summary>
    /// the callback should receive the exact arguments as the function has, and to return the same amount of arguments, plus the first argument needs to be the pointer to the original function
    /// </summary>
    public unsafe static void SetVHook(nint hook, nint entityOrVTable, int index, nint callback, bool isVtable)
    {
        _SetVHook(hook, entityOrVTable, index, callback, isVtable ? (byte)1 : (byte)0);
    }

    private unsafe static delegate* unmanaged<nint, nint, nint, void> _SetMHook;

    /// <summary>
    /// the callback should receive `ref Context64`
    /// </summary>
    public unsafe static void SetMHook(nint hook, nint addr, nint callback)
    {
        _SetMHook(hook, addr, callback);
    }

    private unsafe static delegate* unmanaged<nint, void> _EnableHook;

    public unsafe static void EnableHook(nint hook)
    {
        _EnableHook(hook);
    }

    private unsafe static delegate* unmanaged<nint, void> _EnableVHook;

    public unsafe static void EnableVHook(nint hook)
    {
        _EnableVHook(hook);
    }

    private unsafe static delegate* unmanaged<nint, void> _EnableMHook;

    public unsafe static void EnableMHook(nint hook)
    {
        _EnableMHook(hook);
    }

    private unsafe static delegate* unmanaged<nint, void> _DisableHook;

    public unsafe static void DisableHook(nint hook)
    {
        _DisableHook(hook);
    }

    private unsafe static delegate* unmanaged<nint, void> _DisableVHook;

    public unsafe static void DisableVHook(nint hook)
    {
        _DisableVHook(hook);
    }

    private unsafe static delegate* unmanaged<nint, void> _DisableMHook;

    public unsafe static void DisableMHook(nint hook)
    {
        _DisableMHook(hook);
    }

    private unsafe static delegate* unmanaged<nint, byte> _IsHookEnabled;

    public unsafe static bool IsHookEnabled(nint hook)
    {
        var ret = _IsHookEnabled(hook);
        return ret == 1;
    }

    private unsafe static delegate* unmanaged<nint, byte> _IsVHookEnabled;

    public unsafe static bool IsVHookEnabled(nint hook)
    {
        var ret = _IsVHookEnabled(hook);
        return ret == 1;
    }

    private unsafe static delegate* unmanaged<nint, byte> _IsMHookEnabled;

    public unsafe static bool IsMHookEnabled(nint hook)
    {
        var ret = _IsMHookEnabled(hook);
        return ret == 1;
    }

    private unsafe static delegate* unmanaged<nint, nint> _GetHookOriginal;

    public unsafe static nint GetHookOriginal(nint hook)
    {
        var ret = _GetHookOriginal(hook);
        return ret;
    }

    private unsafe static delegate* unmanaged<nint, nint> _GetVHookOriginal;

    public unsafe static nint GetVHookOriginal(nint hook)
    {
        var ret = _GetVHookOriginal(hook);
        return ret;
    }
}

File: src/scripting/memory/hooks.cpp
/************************************************************************************************
 * SwiftlyS2 is a scripting framework for Source2-based games.
 * Copyright (C) 2023-2026 Swiftly Solution SRL via Sava Andrei-Sebastian and it's contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 ************************************************************************************************/

#include <api/interfaces/manager.h>
#include <cstdio>
#include <scripting/scripting.h>

void* Bridge_Hooks_AllocateHook()
{
    auto hooksmanager = g_ifaceService.FetchInterface<IHooksManager>(HOOKSMANAGER_INTERFACE_VERSION);
    return hooksmanager->CreateFunctionHook();
}

void* Bridge_Hooks_AllocateVHook()
{
    auto hooksmanager = g_ifaceService.FetchInterface<IHooksManager>(HOOKSMANAGER_INTERFACE_VERSION);
    return hooksmanager->CreateVFunctionHook();
}

void* Bridge_Hooks_AllocateMHook()
{
    auto hooksmanager = g_ifaceService.FetchInterface<IHooksManager>(HOOKSMANAGER_INTERFACE_VERSION);
    return hooksmanager->CreateMFunctionHook();
}

void Bridge_Hooks_DeallocateHook(void* hook)
{
    auto hooksmanager = g_ifaceService.FetchInterface<IHooksManager>(HOOKSMANAGER_INTERFACE_VERSION);
    hooksmanager->DestroyFunctionHook((IFunctionHook*)hook);
}

void Bridge_Hooks_DeallocateVHook(void* hook)
{
    auto hooksmanager = g_ifaceService.FetchInterface<IHooksManager>(HOOKSMANAGER_INTERFACE_VERSION);
    hooksmanager->DestroyVFunctionHook((IVFunctionHook*)hook);
}

void Bridge_Hooks_DeallocateMHook(void* hook)
{
    auto hooksmanager = g_ifaceService.FetchInterface<IHooksManager>(HOOKSMANAGER_INTERFACE_VERSION);
    hooksmanager->DestroyMFunctionHook((IMFunctionHook*)hook);
}

void Bridge_Hooks_SetHook(void* hook, void* func, void* callback)
{
    ((IFunctionHook*)hook)->SetHookFunction(func, callback);
}

void Bridge_Hooks_SetVHook(void* hook, void* entityOrVTable, int index, void* callback, bool isVTable)
{
    ((IVFunctionHook*)hook)->SetHookFunction(entityOrVTable, index, callback, isVTable);
}

void Bridge_Hooks_SetMHook(void* hook, void* addr, void* callback)
{
    ((IMFunctionHook*)hook)->SetHookFunction(addr, callback);
}

void Bridge_Hooks_EnableHook(void* hook)
{
    ((IFunctionHook*)hook)->Enable();
}

void Bridge_Hooks_EnableVHook(void* hook)
{
    ((IVFunctionHook*)hook)->Enable();
}

void Bridge_Hooks_EnableMHook(void* hook)
{
    ((IMFunctionHook*)hook)->Enable();
}

void Bridge_Hooks_DisableHook(void* hook)
{
    ((IFunctionHook*)hook)->Disable();
}

void Bridge_Hooks_DisableVHook(void* hook)
{
    ((IVFunctionHook*)hook)->Disable();
}

void Bridge_Hooks_DisableMHook(void* hook)
{
    ((IMFunctionHook*)hook)->Disable();
}

bool Bridge_Hooks_IsHookEnabled(void* hook)
{
    return ((IFunctionHook*)hook)->IsEnabled();
}

bool Bridge_Hooks_IsVHookEnabled(void* hook)
{
    return ((IVFunctionHook*)hook)->IsEnabled();
}

bool Bridge_Hooks_IsMHookEnabled(void* hook)
{
    return ((IMFunctionHook*)hook)->IsEnabled();
}

void* Bridge_Hooks_GetHookOriginal(void* hook)
{
    return ((IFunctionHook*)hook)->GetOriginal();
}

void* Bridge_Hooks_GetVHookOriginal(void* hook)
{
    return ((IVFunctionHook*)hook)->GetOriginal();
}

DEFINE_NATIVE("Hooks.AllocateHook", Bridge_Hooks_AllocateHook);
DEFINE_NATIVE("Hooks.AllocateVHook", Bridge_Hooks_AllocateVHook);
DEFINE_NATIVE("Hooks.AllocateMHook", Bridge_Hooks_AllocateMHook);
DEFINE_NATIVE("Hooks.DeallocateHook", Bridge_Hooks_DeallocateHook);
DEFINE_NATIVE("Hooks.DeallocateVHook", Bridge_Hooks_DeallocateVHook);
DEFINE_NATIVE("Hooks.DeallocateMHook", Bridge_Hooks_DeallocateMHook);
DEFINE_NATIVE("Hooks.SetHook", Bridge_Hooks_SetHook);
DEFINE_NATIVE("Hooks.SetVHook", Bridge_Hooks_SetVHook);
DEFINE_NATIVE("Hooks.SetMHook", Bridge_Hooks_SetMHook);
DEFINE_NATIVE("Hooks.EnableHook", Bridge_Hooks_EnableHook);
DEFINE_NATIVE("Hooks.EnableVHook", Bridge_Hooks_EnableVHook);
DEFINE_NATIVE("Hooks.EnableMHook", Bridge_Hooks_EnableMHook);
DEFINE_NATIVE("Hooks.DisableHook", Bridge_Hooks_DisableHook);
DEFINE_NATIVE("Hooks.DisableVHook", Bridge_Hooks_DisableVHook);
DEFINE_NATIVE("Hooks.DisableMHook", Bridge_Hooks_DisableMHook);
DEFINE_NATIVE("Hooks.IsHookEnabled", Bridge_Hooks_IsHookEnabled);
DEFINE_NATIVE("Hooks.IsVHookEnabled", Bridge_Hooks_IsVHookEnabled);
DEFINE_NATIVE("Hooks.IsMHookEnabled", Bridge_Hooks_IsMHookEnabled);
DEFINE_NATIVE("Hooks.GetHookOriginal", Bridge_Hooks_GetHookOriginal);
DEFINE_NATIVE("Hooks.GetVHookOriginal", Bridge_Hooks_GetVHookOriginal);